File: Striver/DP/Q43.py
from typing import List
import bisect

# longest increasing subsequence

# binary search - Time: O(nlogn), Space: O(n)
def length_of_lis_bin_search(self, arr: List[int], n: int) -> int:
    lis = [arr[0]]
    for i in range(1, n):
        if arr[i] > lis[-1]:
            lis.append(arr[i])
        else:
            ind = bisect.bisect_left(lis, arr[i])
            if ind < 0:
                ind = -ind-1
            lis[ind] = arr[i]
    return len(lis)

File: Striver/DP/Q43.java
package Striver.DP;
import java.util.*;

public class Q43 {
    // longest increasing subsequence

    // binary search - Time: O(nlogn), Space: O(n)
    public int lengthOFLISBinSearch(int[] arr, int n){
        ArrayList<Integer> list = new ArrayList<>();
        list.add(arr[0]);
        for(int i = 1; i < n; i++){
            if(arr[i] > list.get(list.size()-1)){
                list.add(arr[i]);
            }else{
                int ind = Collections.binarySearch(list, arr[i]);
                if(ind < 0) ind = -ind-1;
                list.set(ind, arr[i]);
            }
        }
        return list.size();
    }
}
